packages/gatsby-plugin-image/src/transforms/gatsby-plugin-image.js

```diff
@@ -0,0 +1,283 @@
+import graphql from "gatsby/graphql"
+import { parse, print } from "recast"
+import { transformFromAstSync, parseSync } from "@babel/core"
+
+const legacyFragments = [
+  `GatsbyImageSharpFixed`,
+  `GatsbyImageSharpFixed_withWebp`,
+  `GatsbyImageSharpFluid`,
+  `GatsbyImageSharpFluid_withWebp`,
+]
+
+const legacyFragmentsNoPlaceholder = [
+  `GatsbyImageSharpFixed_noBase64`,
+  `GatsbyImageSharpFixed_withWebp_noBase64`,
+  `GatsbyImageSharpFluid_noBase64`,
+  `GatsbyImageSharpFluid_withWebp_noBase64`,
+]
+
+const legacyFragmentsSVGPlaceholder = [
+  `GatsbyImageSharpFixed_tracedSVG`,
+  `GatsbyImageSharpFixed_withWebp_tracedSVG`,
+  `GatsbyImageSharpFluid_tracedSVG`,
+  `GatsbyImageSharpFluid_withWebp_tracedSVG`,
+]
+
+const typeMapper = {
+  fixed: `FIXED`,
+  fluid: `FLUID`,
+  constrained: `CONSTRAINED`,
+}
+
+export default function jsCodeShift(file) {
+  const transformedSource = babelRecast(file.source, file.path)
+  return transformedSource
+}
+
+export function babelRecast(code, filePath) {
+  const transformedAst = parse(code, {
+    parser: {
+      parse: source =>
+        parseSync(source, {
+          plugins: [
+            `@babel/plugin-syntax-jsx`,
+            `@babel/plugin-proposal-class-properties`,
+          ],
+          overrides: [
+            {
+              test: [`**/*.ts`, `**/*.tsx`],
+              plugins: [[`@babel/plugin-syntax-typescript`, { isTSX: true }]],
+            },
+          ],
+          filename: filePath,
+          parserOpts: {
+            tokens: true, // recast uses this
+          },
+        }),
+    },
+  })
+
+  const options = {
+    cloneInputAst: false,
+    code: false,
+    ast: true,
+    plugins: [updateImport],
+  }
+
+  const { ast } = transformFromAstSync(transformedAst, code, options)
+
+  const result = print(ast, { lineTerminator: `\n` }).code
+  return result
+}
+
+export function updateImport(babel) {
+  const { types: t, template } = babel
+  let imageImportName = ``
+  return {
+    visitor: {
+      ImportDeclaration(path) {
+        const { node } = path
+        if (
+          node.source.value !== `gatsby-image` &&
+          node.source.value !== `gatsby-plugin-image/compat`
+        ) {
+          return
+        }
+        imageImportName = node.specifiers?.[0]?.local?.name
+        const newImport = template.statement
+          .ast`import { GatsbyImage } from "gatsby-plugin-image"`
+        path.replaceWith(newImport)
+      },
+      MemberExpression(path) {
+        if (
+          path.node?.property?.name === `fixed` ||
+          path.node?.property?.name === `fluid`
```

Comment:
Nit: if `path.node` is defined, then `path.node.property` is always defined

```suggestion
          path.node?.property.name === `fixed` ||
          path.node?.property.name === `fluid`
```